ecc/goldilocks/decaf_test.go

```diff
@@ -0,0 +1,137 @@
+package goldilocks_test
+
+import (
+	"bytes"
+	"crypto/rand"
+	"encoding/hex"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/cloudflare/circl/ecc/goldilocks"
+	"github.com/cloudflare/circl/internal/test"
+)
+
+type testJSONFile struct {
+	Group     string `json:"group"`
+	Version   string `json:"version"`
+	Generator struct {
+		X string `json:"x"`
+		Y string `json:"y"`
+		T string `json:"t"`
+		Z string `json:"z"`
+	} `json:"generator"`
+	Vectors []struct {
+		K  string `json:"k"`
+		KG string `json:"kG"`
+		KP string `json:"kP"`
+	} `json:"vectors"`
+}
+
+func (kat *testJSONFile) readFile(t *testing.T, fileName string) {
+	jsonFile, err := os.Open(fileName)
+	if err != nil {
+		t.Fatalf("File %v can not be opened. Error: %v", fileName, err)
+	}
+	defer jsonFile.Close()
+	input, _ := ioutil.ReadAll(jsonFile)
+
+	err = json.Unmarshal(input, &kat)
+	if err != nil {
+		t.Fatalf("File %v can not be loaded. Error: %v", fileName, err)
+	}
+}
+
+func verify(t *testing.T, i int, gotkG *goldilocks.Elt, wantEnckG []byte) {
+	var d goldilocks.Decaf
+	wantkG := &goldilocks.Elt{}
+
+	gotEnckG, err := gotkG.MarshalBinary()
+	got := err == nil && bytes.Equal(gotEnckG, wantEnckG)
+	want := true
+	if got != want {
+		test.ReportError(t, got, want, i)
+	}
+
+	err = wantkG.UnmarshalBinary(wantEnckG)
+	got = err == nil &&
+		d.IsValid(gotkG) &&
+		d.IsValid(wantkG) &&
+		gotkG.IsEqual(wantkG)
+	want = true
+	if got != want {
+		test.ReportError(t, got, want, i)
+	}
+}
+
+func TestDecafv1_1(t *testing.T) {
```

Comment:
You begin with a group element, encode it and then decode it again and see whether it's ok.

You should also check the following: randomly generate a few buffers until it decodes into a group element.  Encode again and check whether you get the same buffer.

Reply:
done

Reply:
:+1: